src/storage/src/vector/mod.rs

```diff
@@ -0,0 +1,537 @@
+        for i in 0..len {
+            // safety: 0 <= i < len
+            unsafe {
+                uninit_mut
+                    .get_unchecked_mut(i)
+                    .write(slice.get_unchecked(i) / magnitude)
+            };
+        }
+        // safety: initialized with len, and have set all item
+        unsafe { VectorInner(uninit.assume_init()) }
+    }
+}
+
+pub type VectorDistance = f32;
+
+pub trait OnNearestItemFn<O> =
+    for<'i> Fn(VectorRef<'i>, VectorDistance, &'i [u8]) -> O + Send + 'static;
+
+#[macro_export]
+macro_rules! for_all_distance_measurement {
+    ($macro:ident $($param:tt)*) => {
+        $macro! {
+            {
+                (L1, $crate::vector::L1Distance),
+                (L2, $crate::vector::L2Distance),
+                (KlDivergence, $crate::vector::KlDivergenceDistance),
+                (Cosine, $crate::vector::CosineDistance),
+                (InnerProduct, $crate::vector::InnerProductDistance),
+            }
+            $($param)*
+        }
+    };
+}
+
+macro_rules! define_measure {
+    ({
+        $(($distance_name:ident, $_distance_type:ty),)+
+    }) => {
+        pub enum DistanceMeasurement {
+            $($distance_name),+
+        }
+    };
+    () => {
+        for_all_distance_measurement! {define_measure}
+    };
+}
+
+define_measure!();
+
+#[macro_export]
+macro_rules! dispatch_measurement {
+    ({
+        $(($distance_name:ident, $distance_type:ty),)+
+    },
+    $measurement:expr, $type_name:ident, $body:expr) => {
+        match $measurement {
+            $(
+                DistanceMeasurement::$distance_name => {
+                    type $type_name = $distance_type;
+                    $body
+                }
+            ),+
+        }
+    };
+    ($measurement:expr, $type_name:ident, $body:expr) => {
+        $crate::for_all_distance_measurement! {dispatch_measurement, $measurement, $type_name, $body}
+    };
+}
+
+pub trait MeasureDistanceBuilder {
+    type Measure<'a>: MeasureDistance<'a>;
+    fn new(target: VectorRef<'_>) -> Self::Measure<'_>;
+
+    #[cfg(test)]
+    fn distance(target: VectorRef<'_>, other: VectorRef<'_>) -> VectorDistance
+    where
+        Self: Sized,
+    {
+        Self::new(target).measure(other)
+    }
+}
+
+pub struct L1Distance;
+
+pub struct L1DistanceMeasure<'a>(VectorRef<'a>);
+
+impl MeasureDistanceBuilder for L1Distance {
+    type Measure<'a> = L1DistanceMeasure<'a>;
+
+    fn new(target: VectorRef<'_>) -> Self::Measure<'_> {
+        L1DistanceMeasure(target)
+    }
+}
+
+impl<'a> MeasureDistance<'a> for L1DistanceMeasure<'a> {
+    fn measure(&self, other: VectorRef<'_>) -> VectorDistance {
+        // TODO: use some library with simd support
+        let len = self.0.0.len();
+        assert_eq!(len, other.0.len());
+        // In this implementation, we don't take the square root to avoid unnecessary computation, because
+        // we only want comparison rather than the actual distance.
```

Comment:
It seems that for L1, even the final result doesn’t require taking the square root.